test/functional/api/autocmd_spec.lua

```diff
@@ -195,33 +197,29 @@ describe('autocmd api', function()
     end)
 
     it('removes an autocommand if the callback returns true', function()
+      local group_name = "ToBeRemoved"
+      meths.set_var("group_name", group_name)
+
       meths.set_var("some_condition", false)
 
-      exec_lua [[
-      vim.api.nvim_create_autocmd("User", {
-        pattern = "Test",
-        desc = "A test autocommand",
-        callback = function()
-          return vim.g.some_condition
-        end,
-      })
+      local result = exec_lua [[
+        local gid = vim.api.nvim_create_augroup(vim.g.group_name, {})
+        local id = vim.api.nvim_create_autocmd("User", {
+          group = vim.g.group_name,
+          pattern = "RemovalTest",
+          desc = "A test autocommand",
+          callback = function()
+            return vim.g.some_condition
+          end,
+        })
+        vim.g.some_condition = true
+        vim.api.nvim_exec_autocmds("User", {pattern = "RemovalTest"})
```

Comment:
Still need to check that the autocommand still exists after it runs and the callback returns `false` (like the existing test does).

Reply:
commenting out the call to `nvim_get_autocmds()` will get the test-case to pass.  it looks like a race-condition between `nvim_get_autocmds()` and `au_cleanup()` (or whatever is responsible for clearing the autocmds)
if this is actually the case, then I'm afraid that the solution won't be trivial

the first thing that jumps to mind is that we'd need to mark the `autocmd` for deletion individually, instead of marking the entire state as "not clean", but that means a large-ish refactor for the logic of clearing the autocmds. see 
https://github.com/neovim/neovim/blob/2c72dc39d718955ac485a08eecd8407641b98599/src/nvim/autocmd.c#L96-L94

Reply:
as discussed, we'll handle this in a separate PR, see #18642